Multiconf

boyleconnor opened the pull request:
Allow Mult-configs as well as Single-configs

boyleconnor committed:
Switch to Luigi Local Scheduler (for Docker)

Use Luigi's local scheduler because we now are using Shilad's Docker
application for the Cartograph server, and we haven't yet worked out how
to make that work with the Luigi central scheduler.

boyleconnor committed:
Change to Use Multi-map Metacofigs

At least one bug remaining: when it updates map_services, it starts a
new MapService for each line in the meta-config, when it should really
just start a new one for the ones that haven't been started yet.

boyleconnor committed:
Prevent Duplicate Map Configs

NewMapService now raises an error when you try to add a map where there
is already a map config file with the same name.

boyleconnor committed:
Prevent Adding Maps with Same Name as any Active Map

boyleconnor committed:
Merge branch 'api' of https://github.com/shilad/cartograph into api

boyleconnor committed:
Undo Erroneous Change in Merge

boyleconnor committed:
Merge branch 'develop' into api

boyleconnor committed:
Refactor into gen_data Function

boyleconnor committed:
Refactor into gen_config Function

boyleconnor committed:
Change gen_data() to Take a List and Return Bad Articles

boyleconnor committed:
Only Update Maps that Need to be Updated

boyleconnor committed:
Refactor previous

boyleconnor committed:
Edit Docstring

boyleconnor committed:
Allow Multi-Config and Single-Config

The server can now be passed by passing it either the path to: a
single cartograph map configuration file *or* a file starting with the
line '######' (no quotes) and whose following lines are all paths to
cartograph map configuration files.

boyleconnor committed:
Merge remote-tracking branch 'remotes/origin/develop' into multiconf

shilad commented:
Would you please pull from develop to make sure you're up to date and re-issue this PR? I'm closing this one so we can have a clean start.

boyleconnor committed:
Merge remote-tracking branch 'remotes/origin/develop' into multiconf

shilad commented on bin/docker-web.sh:
```diff
@@ -23,23 +23,24 @@ function die() {
-# TODO: Switch to multi-process gunicorn, but adding maps dynamically
-# will require cross-process coordination.
-#
 #docker run \
```
Remove commented out code and TODO. They are done! :)

shilad commented on cartograph/server/NewMapService.py:
```diff
@@ -51,6 +51,69 @@ def filter_tsv(source_dir, target_dir, ids, filename):
+    """
+    # Prevent map names with special characters (for security/prevents shell injection)
+    for c in map_name:
+        assert c in ACCEPTABLE_MAP_NAME_CHARS
```
Also allow numbers, spaces, underscores.

shilad commented on cartograph/server/NewMapService.py:
```diff
@@ -51,6 +51,69 @@ def filter_tsv(source_dir, target_dir, ids, filename):
+        os.makedirs(USER_CONF_DIR)
+
+    # Generate a new conf file
+    with open('./data/conf_template.txt', 'r') as conf_template_file:
```
Could you move conf_template.txt into data/conf?

shilad commented on cartograph/server/NewMapService.py:
```diff
@@ -51,6 +51,69 @@ def filter_tsv(source_dir, target_dir, ids, filename):
                 popularities_writer.writerow(row)
 
 
```
Feature request: Could you write a function that checks to see whether a map name is valid (characters, does not exist) and returns an error message or None?

shilad commented on cartograph/server/NewMapService.py:
```diff
@@ -51,6 +51,69 @@ def filter_tsv(source_dir, target_dir, ids, filename):
+    """
+    # Generate dictionary of article names to IDs
+    # TODO: is there a way to do this once (instead of once per POST)?
+    names_path = os.path.join(SOURCE_DIR, 'names.tsv')
```
Use pandas for all I/O.

shilad commented on cartograph/server/NewMapService.py:
```diff
@@ -51,6 +51,69 @@ def filter_tsv(source_dir, target_dir, ids, filename):
+    ids = []
+    bad_articles = []
+    for term in articles:
+        try:
```
In general it's better use the "in" operator and avoid KeyErrors.

boyleconnor replied:
In the other branch (api-metrics) ids is a set(), so let's wait until the next branch gets added in

shilad commented on cartograph/server/NewMapService.py:
```diff
@@ -51,6 +51,69 @@ def filter_tsv(source_dir, target_dir, ids, filename):
+    bad_articles = []
+    for term in articles:
+        try:
+            ids += [name_dict[term]]  # Attempts to find entry in dict of Articles to IDs
```
use ids.append to improve performance. It's O(1) vs O(n)

boyleconnor replied:
In the other branch (api-metrics) ids is a set(), so I'm not gonna add this in

shilad commented on cartograph/server/NewMapService.py:
```diff
@@ -61,55 +124,31 @@ def on_get(self, req, resp):
-
-        # Build from the new conf file
-        os.system("CARTOGRAPH_CONF=""%s"" PYTHONPATH=$PYTHONPATH:.:./cartograph luigi --module cartograph ParentTask" % config_path)
+        articles = post_data['articles'].split('\r\n')
```
Support either line ending ('\r\n' or '\r'): ``.split("[\\r\\n]+")``

boyleconnor replied:
It looks like str().split might not support regex (at least it doesn't seem to for me) so I'm using re.split

shilad commented on cartograph/server/NewMapService.py:
```diff
@@ -61,55 +124,31 @@ def on_get(self, req, resp):
+        config_path = gen_config(map_name)
+
+        # Build from the new config file
+        os.system("CARTOGRAPH_CONF=""%s"" PYTHONPATH=$PYTHONPATH:.:./cartograph luigi --module cartograph ParentTask --local-scheduler" % config_path)
```
Check exit code... I think you'll need to use subprocess instead. Also (if possible) it would be ideal for the build log to go to a file in the map directory.

shilad commented on cartograph/server/ParentService.py:
```diff
@@ -23,13 +30,41 @@ def service_for_map(self, map_name):
+        :return:
+        """
+        with open(meta_config, 'r') as configs:
+            configs.next()  # Skip the multi-map flag (i.e. '######')
```
assert that it is the right line

shilad commented on cartograph/server/ParentService.py:
```diff
@@ -23,13 +30,41 @@ def service_for_map(self, map_name):
+                    self.map_services[map_service.name] = map_service
+
+        # indicate that map_services has been updated
+        self.map_services['_last_update'] = os.stat(meta_config)
```
Use mtime instead of all of os.stat. Os.stat includes access time, which may be updated without any real changes.

shilad reviewed (changes requested):
Very nice overall! The codebase is so much more readable. I feel happy! A few small changes.

boyleconnor committed:
Add Comments Describing map_services

map_services is a weird, singleton thing to which most objects in this
project get passed pointers. This commit adds some comments
clarifying that, particulary in ParentService.

boyleconnor committed:
Add Underscores and Digits to Acceptable Map Name Chars

boyleconnor committed:
Remove Completed Gunicorn TODO

boyleconnor committed:
Adding Back in Original Test Config File

I overwrote this and added it to Git on accident. Whoops! It happens to
be necessary for unit tests.

boyleconnor committed:
Refactor Map Name Checking

boyleconnor committed:
Split Lines in a Platform-Independently

boyleconnor committed:
Switch Luigi Build Command from OS to Subprocess

boyleconnor committed:
Pipe Build Output into Files

boyleconnor committed:
Double-check for Metaconf Flag

This also includes a refactoring of how the map config paths are loaded
from the metaconf file.

boyleconnor committed:
Switch From os.stat() to getmtime()

os.stat apparently changes even when a file is read, and we only want to
update the server when the meta-config is actually *modified*

shilad reviewed (approved):
Very nice! LGTM
